# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: toolkit/components/antitracking/bouncetrackingprotection/BounceTrackingState.cpp
# Commit: 087b27593538
# Full Hash: 087b27593538ac10d0c69e41100423f430ad8c89
# Author: Paul Zuehlcke <pbz@mozilla.com>
# Date: 2023-09-07 21:34:02
# Regressor Bug: 1839918
# File Overlap Count: 1
# Description:
#   Bug 1839918 - Add BounceTrackingStorageObserver to observe cookie changes. r=bvandersloot,anti-tracking-reviewers
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D185304
# ==============================================================================

diff -r 302d12c12083 -r 087b27593538 toolkit/components/antitracking/bouncetrackingprotection/BounceTrackingState.cpp
--- a/toolkit/components/antitracking/bouncetrackingprotection/BounceTrackingState.cpp	Thu Sep 07 08:25:55 2023 +0000
+++ b/toolkit/components/antitracking/bouncetrackingprotection/BounceTrackingState.cpp	Thu Sep 07 08:25:56 2023 +0000
@@ -8,12 +8,14 @@
 #include "BounceTrackingState.h"
 #include "BounceTrackingRecord.h"
 
+#include "BounceTrackingStorageObserver.h"
 #include "ErrorList.h"
 #include "mozilla/dom/BrowsingContext.h"
 #include "mozilla/dom/BrowsingContextWebProgress.h"
 #include "mozilla/dom/CanonicalBrowsingContext.h"
 #include "nsCOMPtr.h"
 #include "nsDebug.h"
+#include "nsError.h"
 #include "nsIBrowser.h"
 #include "nsIChannel.h"
 #include "nsIEffectiveTLDService.h"
@@ -33,6 +35,8 @@
 static StaticAutoPtr<nsTHashMap<uint64_t, RefPtr<BounceTrackingState>>>
     sBounceTrackingStates;
 
+static StaticRefPtr<BounceTrackingStorageObserver> sStorageObserver;
+
 NS_IMPL_ISUPPORTS(BounceTrackingState, nsIWebProgressListener,
                   nsISupportsWeakReference);
 
@@ -64,6 +68,14 @@
     ClearOnShutdown(&sBounceTrackingStates);
   }
 
+  if (!sStorageObserver) {
+    sStorageObserver = new BounceTrackingStorageObserver();
+    ClearOnShutdown(&sStorageObserver);
+
+    DebugOnly<nsresult> rv = sStorageObserver->Init();
+    NS_WARNING_ASSERTION(NS_SUCCEEDED(rv), "Failed to init storage observer");
+  }
+
   dom::BrowsingContext* browsingContext = aWebProgress->GetBrowsingContext();
   if (!browsingContext) {
     return nullptr;
@@ -524,4 +536,19 @@
   return NS_OK;
 }
 
+nsresult BounceTrackingState::OnCookieWrite(const nsACString& aSiteHost) {
+  NS_ENSURE_TRUE(!aSiteHost.IsEmpty(), NS_ERROR_FAILURE);
+
+  MOZ_LOG(gBounceTrackingProtectionLog, LogLevel::Verbose,
+          ("%s: OnCookieWrite: %s.", __FUNCTION__,
+           PromiseFlatCString(aSiteHost).get()));
+
+  if (!mBounceTrackingRecord) {
+    return NS_OK;
+  }
+
+  mBounceTrackingRecord->AddStorageAccessHost(aSiteHost);
+  return NS_OK;
+}
+
 }  // namespace mozilla